# Notes to Improve Workflow

In [53]:
# Setup matplotlib to plot inline (within the notebook)
%matplotlib inline
# Import the pyplot module of Matplotlib as plt
import matplotlib.pyplot as plt
# Import pandas under the abbreviation 'pd'
import pandas as pd
# Import NumPy under the abbreviation 'np'
import numpy as np
# Libraries to aid in web scraping
import requests
from bs4 import BeautifulSoup

#Notes for efficienccy:
#shift + tab when cursor is in arguments to bring up its documentation
#ex: pd.DataFrame(Shift+tab here)
#recall: Json == Dictionary
#to know functions of pandas:
dir(pd)

['BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_hashtable',
 '_is_numpy_dev',
 '_lib',
 '_libs',
 '_np_version_under1p18',
 '_testing',
 '_tslib',
 '_typing',
 

# Reading The Data

In [2]:
# using pandas and read_csv()
acquis = pd.read_csv("datasets/startups/acquisitions.csv")
additions = pd.read_csv("datasets/startups/additions.csv")
companies = pd.read_csv("datasets/startups/companies.csv")
invests = pd.read_csv("datasets/startups/investments.csv")
rounds = pd.read_csv("datasets/startups/rounds.csv")
dfAll = [acquis, additions, companies, invests, rounds]

In [3]:
#shapes (rows x cols) of each dataset
for ele in dfAll:
    print(ele.shape)

(18969, 18)
(2213, 4)
(66368, 14)
(168647, 18)
(114949, 12)


# Samples of the datasets

In [4]:
acquis.sample(5).transpose()

16705  \
company_permalink            /organization/the-kelsey-group   
company_name                               The Kelsey Group   
company_category_list                                   NaN   
company_country_code                                    NaN   
company_state_code                                      NaN   
company_region                                          NaN   
company_city                                            NaN   
acquirer_permalink      /organization/bia-financial-network   
acquirer_name                         BIA Financial Network   
acquirer_category_list                           Consulting   
acquirer_country_code                                   USA   
acquirer_state_code                                      VA   
acquirer_region                            Washington, D.C.   
acquirer_city                                     Chantilly   
acquired_at                                       10/6/2008   
acquired_month                                      2008-10   
price_amount                                            NaN   
price_currency_code                                     USD   

                                                                   11508  \
company_permalink                                /organization/nexplanar   
company_name                                                   NexPlanar   
company_category_list                                     Semiconductors   
company_country_code                                                 USA   
company_state_code                                                    OR   
company_region                                          Portland, Oregon   
company_city                                                   Hillsboro   
acquirer_permalink      /organization/cabot-microelectronics-corporation   
acquirer_name                         Cabot Microelectronics Corporation   
acquirer_category_list                                    Semiconductors   
acquirer_country_code                                                USA   
acquirer_state_code                                                   IL   
acquirer_region                                                  Chicago   
acquirer_city                                                     Aurora   
acquired_at                                                   12/28/2015   
acquired_month                                                   2015-12   
price_amount                                                         NaN   
price_currency_code                                                  USD   

                                               15222  \
company_permalink       /organization/skywave-mobile   
company_name                          SkyWave Mobile   
company_category_list            Gps|Mobile|Tracking   
company_country_code                             CAN   
company_state_code                                ON   
company_region                                Ottawa   
company_city                                  Ottawa   
acquirer_permalink             /organization/orbcomm   
acquirer_name                                ORBCOMM   
acquirer_category_list            Telecommunications   
acquirer_country_code                            USA   
acquirer_state_code                               NJ   
acquirer_region                               Newark   
acquirer_city                          Rochelle Park   
acquired_at                                11/6/2014   
acquired_month                               2014-11   
price_amount                             130000000.0   
price_currency_code                              USD   

                                                                    16091  \
company_permalink                          /organization/superderivatives   
company_name                                             SuperDerivatives   
company_category_list   Analytics|Finance|FinTech|Information Services...   
company_country_code                  

In [5]:
additions.sample(5).transpose()

988           152      313           177   \
content    Profiles: People  Acquisitions     IPOs  Acquisitions   
month_str           2011-01       2003-11  1979-01       2005-12   
year_str               2011          2003     1979          2005   
value                  2403            54        2            57   

                       1038  
content    Profiles: People  
month_str           2015-03  
year_str               2015  
value                 12945

In [6]:
companies.sample(5).transpose()

47891  \
permalink                          /organization/repica   
name                                             repica   
homepage_url                      http://www.repica.jp/   
category_list      Email|Gift Card|Subscription Service   
funding_total_usd                                874789   
status                                        operating   
country_code                                        NaN   
state_code                                          NaN   
region                                              NaN   
city                                                NaN   
funding_rounds                                        1   
founded_at                                     8/1/2006   
first_funding_at                              11/6/2014   
last_funding_at                               11/6/2014   

                                   66157  \
permalink           /organization/zomazz   
name                              Zomazz   
homepage_url       http://www.zomazz.com   
category_list                Curated Web   
funding_total_usd                2040342   
status                         operating   
country_code                         USA   
state_code                            CA   
region                       Los Angeles   
city                       Monterey Park   
funding_rounds                         1   
founded_at                      1/1/2010   
first_funding_at               6/20/2012   
last_funding_at                6/20/2012   

                                                    46533  \
permalink          /organization/quintessence-biosciences   
name                             Quintessence Biosciences   
homepage_url                          http://quintbio.com   
category_list                               Biotechnology   
funding_total_usd                                 1486000   
status                                             closed   
country_code                                          USA   
state_code                                             WI   
region                                            Madison   
city                                              Madison   
funding_rounds                                          1   
founded_at                                       1/1/2000   
first_funding_at                                6/26/2012   
last_funding_at                                 6/26/2012   

                                       34640  \
permalink                  /organization/maz   
name                                     MAZ   
homepage_url       http://www.mazdigital.com   
category_list         iPad|Mobile|Publishing   
funding_total_usd                    1000000   
status                             operating   
country_code                             USA   
state_code                                NY   
region                         New York City   
city                                New York   
funding_rounds                             3   
founded_at                          7/1/2010   
first_funding_at                  11/14/2012   
last_funding_at                     8/1/2013   

                                                               24481  
permalink                                    /organization/gura-gear  
name                                                       Gura Gear  
homepage_url                                    http://guragear.com/  
category_list      Local Businesses|Professional Services|Retail|...  
funding_total_usd                                            1300000  
status                                                     operating  
country_code                                                     USA  
state_code                                                        UT  
region                                                Salt Lake City  
city                                                           Ogden  
funding_rounds                                                     1  
founded_at    

In [7]:
invests.sample(5).transpose()

93360   \
company_permalink                            /organization/motionmetrics   
company_name                                               MotionMetrics   
company_category_list                      Hardware + Software|Wearables   
company_country_code                                                 GBR   
company_state_code                                                    H9   
company_region                                                    London   
company_city                                                      London   
investor_permalink                              /person/sean-osullivan-2   
investor_name                                            Sean O'Sullivan   
investor_country_code                                                USA   
investor_state_code                                                  NaN   
investor_region                                                      NaN   
investor_city                                                   New York   
funding_round_permalink  /funding-round/3b61482b61fc47a9bdd435c6b2ba323a   
funding_round_type                                                 grant   
funding_round_code                                                   NaN   
funded_at                                                       9/1/2014   
raised_amount_usd                                                50000.0   

                                                                  90852   \
company_permalink                         /organization/mirabilis-medica   
company_name                                            Mirabilis Medica   
company_category_list                                        Health Care   
company_country_code                                                 USA   
company_state_code                                                    WA   
company_region                                                   Seattle   
company_city                                                     Bothell   
investor_permalink                   /organization/charter-life-sciences   
investor_name                                      Charter Life Sciences   
investor_country_code                                                USA   
investor_state_code                                                   CA   
investor_region                                              SF Bay Area   
investor_city                                                Santa Clara   
funding_round_permalink  /funding-round/c917520a9e6409561c89f0752de5b2bd   
funding_round_type                                               venture   
funding_round_code                                                     A   
funded_at                                                      8/22/2007   
raised_amount_usd                                             10500000.0   

                                                                  16295   \
company_permalink                             /organization/beauty-noted   
company_name                                                Beauty Noted   
company_category_list                                             Beauty   
company_country_code                                                 USA   
company_state_code                                                    CA   
company_region                                               SF Bay Area   
company_city                                                  Menlo Park   
investor_permalink               /organization/kortschak-investments-l-p   
investor_name                                Kortschak Investments, L.P.   
investor_country_code                                                USA   
investor_state_code                                                   CA   
investor_region                                              SF Bay Area   
investor_city                                                 Menlo Park   
funding_round_permalink  /funding-round/b9dd43c96e9083bdaac8442b79687d63   
funding_round_type                  

In [8]:
rounds.sample(5).transpose()

90830  \
company_permalink                         /organization/slips-technologies   
company_name                                            SLIPS Technologies   
company_category_list    Assisitive Technology|Chemicals|Specialty Chem...   
company_country_code                                                   USA   
company_state_code                                                      MA   
company_region                                                      Boston   
company_city                                                     Cambridge   
funding_round_permalink    /funding-round/640110efcf6827953852a32762c5d6f4   
funding_round_type                                                 venture   
funding_round_code                                                       A   
funded_at                                                       10/29/2014   
raised_amount_usd                                                3000000.0   

                                                                   43287  \
company_permalink                   /organization/health-benefits-direct   
company_name                                      Health Benefits Direct   
company_category_list                                    Finance|FinTech   
company_country_code                                                 USA   
company_state_code                                                    PA   
company_region                                              Philadelphia   
company_city                                                      Radnor   
funding_round_permalink  /funding-round/9405d966743e2cdea694b52dca82fe19   
funding_round_type                                               venture   
funding_round_code                                                   NaN   
funded_at                                                      10/1/2010   
raised_amount_usd                                              5400000.0   

                                                                   62882  \
company_permalink                            /organization/mobile-patrol   
company_name                                               Mobile Patrol   
company_category_list                                           Software   
company_country_code                                                 NaN   
company_state_code                                                   NaN   
company_region                                                       NaN   
company_city                                                         NaN   
funding_round_permalink  /funding-round/050bb45b451cbc7797f20f49c924476f   
funding_round_type                                               venture   
funding_round_code                                                   NaN   
funded_at                                                     10/14/2011   
raised_amount_usd                                               250000.0   

                                                                   98556  \
company_permalink                                  /organization/telderi   
company_name                                                     Telderi   
company_category_list                                                NaN   
company_country_code                                                 NaN   
company_state_code                                                   NaN   
company_region                                                       NaN   
company_city                                                         NaN   
funding_round_permalink  /funding-round/367397d88b4bddb42af2ec227fb31270   
funding_round_type                                               venture   
funding_round_code                                                   NaN   
funded_at                                                       1/1/2013   
raised_amount_usd                                                    NaN   

                                                                   54789  
company_perm

# Seeing the columns in the datasets

In [9]:
def datasetsCols(dfList, datasetNames, pad=""):
    """
    Takes dataframes and returns a dataframe (df) with each column having a df's
    column names. The "pad" is to make sure all lists of columns have 
    the same length padded out with "pad"
    Example
    pad = "XX", datasetNames = ["cars", "planes"]
    dfList = [carsDF, planesDF]
    returned data frame:
            cars            planes
    0       numOfWheels     numOfWings
    1       manufacturer    manufacturer
    2       make            XX
    """
    cols = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        dictCsvs.update({datasetNames[i] : cols[i]})

    return pd.DataFrame(dictCsvs)

In [10]:
lstCsvNames = ["acquisitions", "additions", "companies", "investments", "rounds"]

dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions          companies  \
0        company_permalink    content          permalink   
1             company_name  month_str               name   
2    company_category_list   year_str       homepage_url   
3     company_country_code      value      category_list   
4       company_state_code             funding_total_usd   
5           company_region                        status   
6             company_city                  country_code   
7       acquirer_permalink                    state_code   
8            acquirer_name                        region   
9   acquirer_category_list                          city   
10   acquirer_country_code                funding_rounds   
11     acquirer_state_code                    founded_at   
12         acquirer_region              first_funding_at   
13           acquirer_city               last_funding_at   
14             acquired_at                                 
15          acquired_month                                 
16            price_amount                                 
17     price_currency_code                                 

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

# Data Cleaning

First, lets change the columns in the `companies` dataset <br>
to be the similar to all other datasets: <br>
(we're doing this in case we merge the datasets together, we want them to have the same column names)

In [11]:
companies.rename(columns = {'permalink' : 'company_permalink', 
                            'name' : 'company_name',
                            'category_list' : 'company_category_list',
                            'country_code' : 'company_country_code',
                            'state_code' : 'company_state_code',
                            'region' : 'company_region',
                            'city' : 'company_city'}, inplace = True)

In [12]:
dfAll[2] = companies
dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

Let's start by analyzing `additions` dataset, as it has unusual format of columns

## Removing "additions" Dataset

In [13]:
additions.sample(5)

content month_str  year_str  value
283                  Acquisitions   2014-10      2014    290
1545  Rounds: equity_crowdfunding   2014-11      2014    153
1785       Rounds: private_equity   2010-12      2010      3
1182                       Rounds   2009-11      2009    433
1482       Rounds: debt_financing   2013-05      2013    135

In [14]:
additions["content"].unique()

array(['Acquisitions', 'IPOs', 'Profiles', 'Profiles: Organizations',
       'Profiles: People', 'Profiles: Products', 'Rounds',
       'Rounds: angel', 'Rounds: convertible_note',
       'Rounds: debt_financing', 'Rounds: equity_crowdfunding',
       'Rounds: grant', 'Rounds: non_equity_assistance',
       'Rounds: post_ipo_debt', 'Rounds: post_ipo_equity',
       'Rounds: private_equity', 'Rounds: product_crowdfunding',
       'Rounds: secondary_market', 'Rounds: seed', 'Rounds: undisclosed',
       'Rounds: venture'], dtype=object)

In [15]:
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

In [16]:
additions.shape

(2213, 4)

In [17]:
companies.shape, rounds.shape

((66368, 14), (114949, 12))

So apparently there are no columns that could be used to join with the other datasets. <br>
Furthermore, the "value" column is too vague to be useful. <br>
Therefore, the "additions" dataset will be discarded.

In [18]:
dfAll.pop(1)
lstCsvNames.pop(1)
len(dfAll), lstCsvNames

(4, ['acquisitions', 'companies', 'investments', 'rounds'])

# Data Cleaning (Cont.)

## Checking All Missing Values

In [19]:
def datasetsNulls(dfList, datasetNames, nullCol="nulls_", pad=""):
    """
    Use this when you want to display a column of column names,
    then a column of the null values, and repeat that for each dataset.
    Returns a dataframe
    """
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.isnull().sum().tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[nullCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [20]:
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        1           company_name      1.0   
2    company_category_list     4984           homepage_url   5058.0   
3     company_country_code     4669  company_category_list   3148.0   
4       company_state_code     4949      funding_total_usd      0.0   
5           company_region     4911                 status      0.0   
6             company_city     4911   company_country_code   6958.0   
7       acquirer_permalink        1     company_state_code   8547.0   
8            acquirer_name        1         company_region   8030.0   
9   acquirer_category_list     2389           company_city   8028.0   
10   acquirer_country_code     1171         funding_rounds      0.0   
11     acquirer_state_code     1436             founded_at  15221.0   
12         acquirer_region     1360       first_funding_at     24.0   
13           acquirer_city     1360        last_funding_at      0.0   
14             acquired_at       13                             NaN   
15          acquired_month       15                             NaN   
16            price_amount    13957                             NaN   
17     price_currency_code        7                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        1             company_name      1.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink       12  funding_round_permalink      0.0  
8             investor_name       12       funding_round_type      0.0  
9     investor_country_code    30939       funding_round_code  83809.0  
10      investor_state_code    44739                funded_at      0.0  
11          investor_region    43217        raised_amount_usd  19990.0  
12            investor_city    32164                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87585                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

### `dropna()` for columns with few missing values

Notice that there is one row that doesn't contain a company name, <br>
so let's validate that it is the same company across all datasets, <br>
so we can remove it:

In [21]:
datasetsCols(dfAll, lstCsvNames)

acquisitions              companies              investments  \
0        company_permalink      company_permalink        company_permalink   
1             company_name           company_name             company_name   
2    company_category_list           homepage_url    company_category_list   
3     company_country_code  company_category_list     company_country_code   
4       company_state_code      funding_total_usd       company_state_code   
5           company_region                 status           company_region   
6             company_city   company_country_code             company_city   
7       acquirer_permalink     company_state_code       investor_permalink   
8            acquirer_name         company_region            investor_name   
9   acquirer_category_list           company_city    investor_country_code   
10   acquirer_country_code         funding_rounds      investor_state_code   
11     acquirer_state_code             founded_at          investor_region   
12         acquirer_region       first_funding_at            investor_city   
13           acquirer_city        last_funding_at  funding_round_permalink   
14             acquired_at                              funding_round_type   
15          acquired_month                              funding_round_code   
16            price_amount                                       funded_at   
17     price_currency_code                               raised_amount_usd   

                     rounds  
0         company_permalink  
1              company_name  
2     company_category_list  
3      company_country_code  
4        company_state_code  
5            company_region  
6              company_city  
7   funding_round_permalink  
8        funding_round_type  
9        funding_round_code  
10                funded_at  
11        raised_amount_usd  
12                           
13                           
14                           
15                           
16                           
17

In [22]:
#company_permalink of each row that doesn't have a company name:
for df in dfAll:
    print(df[df['company_name'].isnull()]['company_permalink'].values)

['mendatio"']
['/organization/tell-it-in']
['/organization/tell-it-in']
['/organization/tell-it-in']


So apparently, the `acquisitions` dataset has a different permalink that has no company name, <br>
So we could remove these two companies, as the total number of companies are big enough to get the insight that we want.

`dfAllNulls` dataframe also shows that there are missing values for `acquirer_name` and `investor_name` <br>
which can't be imputed, so we'll also remove them.

Finally, we'll remove the missing values for columns that have less than 100 `NaN`, <br>
as they're small in comparison to their respective datasets:

In [23]:
for df in dfAll:
    nulls = df.isnull().sum()
    cols = nulls[(nulls <= 100) & (nulls != 0)].index.tolist()
    df.dropna(subset=cols, inplace=True)
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        0           company_name      0.0   
2    company_category_list     4977           homepage_url   5058.0   
3     company_country_code     4661  company_category_list   3147.0   
4       company_state_code     4940      funding_total_usd      0.0   
5           company_region     4902                 status      0.0   
6             company_city     4902   company_country_code   6956.0   
7       acquirer_permalink        0     company_state_code   8545.0   
8            acquirer_name        0         company_region   8028.0   
9   acquirer_category_list     2387           company_city   8026.0   
10   acquirer_country_code     1168         funding_rounds      0.0   
11     acquirer_state_code     1432             founded_at  15218.0   
12         acquirer_region     1356       first_funding_at      0.0   
13           acquirer_city     1356        last_funding_at      0.0   
14             acquired_at        0                             NaN   
15          acquired_month        0                             NaN   
16            price_amount    13941                             NaN   
17     price_currency_code        0                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        0             company_name      0.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink        0  funding_round_permalink      0.0  
8             investor_name        0       funding_round_type      0.0  
9     investor_country_code    30927       funding_round_code  83808.0  
10      investor_state_code    44727                funded_at      0.0  
11          investor_region    43205        raised_amount_usd  19990.0  
12            investor_city    32152                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87576                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

## Converting & Imputing Dates

[SAY WHAT YOU WILL DO HERE FARAH:]

In [24]:
#Farah: plz impute (not drop) the columns' missing values related to dates in each dataset which are:
#founded_at.. and that's it :] 
# Suggestion: convert the dates of all the columns that have date values from "object" to datetime
#       Helpful link: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#converting-to-timestamps
# Then, make the founded_at a couple of days before first_funding_at

#Sincerely No. -Farah :)

We are currently interested in 'founded_at', 'first_funding_at' and 'last_funding_at'. Before doing operations on dates, they need to be converted first from 'Object' to 'DateTime'.

In [25]:
companies['founded_at'] =pd.to_datetime(companies['founded_at'], errors = 'coerce')
companies['first_funding_at'] =pd.to_datetime(companies['first_funding_at'], errors = 'coerce')
companies['last_funding_at'] =pd.to_datetime(companies['last_funding_at'], errors = 'coerce')
companies.dtypes

company_permalink                object
company_name                     object
homepage_url                     object
company_category_list            object
funding_total_usd                object
status                           object
company_country_code             object
company_state_code               object
company_region                   object
company_city                     object
funding_rounds                    int64
founded_at               datetime64[ns]
first_funding_at         datetime64[ns]
last_funding_at          datetime64[ns]
dtype: object

After converting, we need to impute the missing values on the dates of when those companies where founded at. So, this will be done by first creating a new column that calculates how long it took to get the first funing. This will be done by subtracting the date of the founding from the date of the first funding.

In [26]:
companies['time_before_first_fund'] = companies['first_funding_at'] - companies['founded_at']
companies['time_before_first_fund']

0             NaT
1       -187 days
2             NaT
3        443 days
4       1665 days
           ...   
66363         NaT
66364   -194 days
66365   1308 days
66366    365 days
66367         NaT
Name: time_before_first_fund, Length: 66343, dtype: timedelta64[ns]

Now we need to calculate the mean of the time before the first funding and impute the missing values with this average.

In [27]:
companies['time_before_first_fund'].fillna(companies['time_before_first_fund'].mean(), inplace = True)
companies['time_before_first_fund']

0       1370 days 03:49:18.015610624
1                -187 days +00:00:00
2       1370 days 03:49:18.015610624
3                  443 days 00:00:00
4                 1665 days 00:00:00
                    ...             
66363   1370 days 03:49:18.015610624
66364            -194 days +00:00:00
66365             1308 days 00:00:00
66366              365 days 00:00:00
66367   1370 days 03:49:18.015610624
Name: time_before_first_fund, Length: 66343, dtype: timedelta64[ns]

After calculating the mean and filling the nulls with it, now we can impute the missing dates in the 'founded_at' column by subtracting the time before first funding from the first funding date.

In [28]:
companies['founded_at'].fillna(companies['first_funding_at'] - companies['time_before_first_fund'], inplace = True)
companies['founded_at']

0       2011-04-05 20:10:41.984389376
1       2014-09-04 00:00:00.000000000
2       2010-04-30 20:10:41.984389376
3       2007-01-01 00:00:00.000000000
4       2010-01-01 00:00:00.000000000
                     ...             
66363   2008-06-30 20:10:41.984389376
66364   2012-05-13 00:00:00.000000000
66365   2011-01-01 00:00:00.000000000
66366   2014-01-01 00:00:00.000000000
66367   2009-12-30 20:10:41.984389376
Name: founded_at, Length: 66343, dtype: datetime64[ns]

In [29]:
#farah stop point here

## Cleaning & Imputing Geographical Locations

### Removing `company_state_code` and `company_region`

We don't need `company_state_code` in our analysis, <br>
As country code and city are sufficient to know <br>
an approximation of the geographical location of the company. <br>
Therefore, let's remove it:

In [30]:
companies[['company_country_code', 'company_state_code', 'company_city']].head(10)

company_country_code company_state_code   company_city
0                  IND                 16         Mumbai
1                  USA                 DE  Delaware City
2                  NaN                NaN            NaN
3                  CHN                 22        Beijing
4                  USA                 IL      Champaign
5                  HKG                NaN      Hong Kong
6                  CAN                 BC      Vancouver
7                  USA                 CA  Mountain View
8                  USA                 CA  San Francisco
9                  CHL                 12     Las Condes

In [31]:
companies.drop('company_state_code', axis=1, inplace=True)

`company_region` is also redundent, as it is usually the same as `company_city`.<br>
Let's validate this:

In [32]:
# Note: same logic applies for the permalinks
acquis.dropna(subset=['acquirer_permalink', 'acquirer_name'], inplace=True)
invests.dropna(subset=['investor_permalink', 'investor_name'], inplace=True)

Also, we'll remove the missing values for columns that have less than 100 `NaN`, <br>
as they're small in comparison to their respective datasets:

In [33]:
companies[companies.company_region == companies.company_city][['company_region', 'company_city']]

company_region company_city
0             Mumbai       Mumbai
3            Beijing      Beijing
5          Hong Kong    Hong Kong
6          Vancouver    Vancouver
11           Orlando      Orlando
...              ...          ...
66354           Bath         Bath
66360      Milwaukee    Milwaukee
66362         London       London
66363        Beijing      Beijing
66364          Split        Split

[25360 rows x 2 columns]

Let's see the mean of similarity between the strings using builtin library `difflib.SequenceMatcher`:

In [34]:
from difflib import SequenceMatcher
similarities = []
rgs = companies.company_region.values.tolist()
cts = companies.company_city.values.tolist()
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]): # If either of the elements is NaN, then don't find the similarity
        similarities.append(SequenceMatcher(None, rgs[i], cts[i]).ratio()) # Gets the similarity between each two strings
avgSim = sum(similarities) / len(similarities)
avgSim

0.6175093624179544

However, `SequenceMatcher` only finds the longest contiguous matching subsequence, <br>
and all other characters will drastically decrease the similarity, which isn't very accurate, <br>
Therefore, let's try to see how many times a string is a substring of another:

In [35]:
subsCount = 0
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]):
        subsCount += (rgs[i] in cts[i] or cts[i] in rgs[i])
subsCount

29884

So to summarize: almost half of the companies have the same data for `region` and `city` (25360) <br>
There are around 5000 companies that have the same semantic meaning between the data <br>
(e.g: 'New York City' is the same as 'NYC, New York City') <br>
Therefore, `company_region` can be removed without losing any possible future insights:

In [36]:
companies.drop('company_region', axis=1, inplace=True)

### Imputing `company_country_code` and `company_city`

Steps to impute country codes is to do the following for all missing values:
1. check the company's permalink
2. find HTML that contains the headquarters location (country and city)
3. scrape country and convert it to country code using `pycountry.countries`
4. scrape city and impute it in `company_city`

Step 1: appending "https://www.crunchbase.com" to each permalink <br>
to be able to access the company's webpage on crunchbase. <br>
However, since the new permalink will be much longer (thus slower for accessing as a dataset's primary index) <br>
we'll create a function that appends the string as a prefix and that will be accessed throughout the notebook

In [52]:
##################################### function "visitLink()" used throughout the notebook #####################################
visitLink = lambda s : "https://www.crunchbase.com" + s

Step 2: Finding location of country and city on the webpage:
let's check company [004](https://www.crunchbase.com/organization/004)'s webpage for example: <Br>
<img src="Phase 1/countryAndCityInCrunchbase.png" width=400 height=300 />

By inspecting the HTML, we notice 4 instances of the class <br>
`component--field-formatter field-type-identifier-multi` (in a `span` element),<br>and the data that we want
is always the first instance: <br>
<img src="Phase 1/countryAndCityLocationInHTML.png" width=700 height=500 />

Step 3: scrape that info using `requests` and `bs4` libraries: <br>
(Note, we're using `threading.Thread` here to increase performance by scraping in multithreads)

In [56]:
from threading import Thread
def imputeFromLinks(df, linkCol, colToImpute, linkPrefix = ""):
    links = df[df[colToImpute].isnull()][[linkCol]] # leave it as a series (i.e. don't ".tolist()") to impute it in fillna() below
    results = [Thread(target=fetchCountry, args=(linkPrefix+link)) for link in links]
    

36

In [37]:
companies[companies.company_country_code.isnull() == False].sample(10)

company_permalink  \
55429                     /organization/supercell   
49660                   /organization/savemeeting   
20354                   /organization/filtosh-inc   
46026                        /organization/putney   
30774                     /organization/kite-ping   
61057                 /organization/valor-medical   
57516   /organization/the-little-blue-book-mobile   
9372                        /organization/camiant   
26136                     /organization/huitongda   
32206  /organization/librestream-technologies-inc   

                        company_name                  homepage_url  \
55429                      Supercell         http://supercell.com/   
49660                    SaveMeeting    http://www.savemeeting.com   
20354                   Filtosh Inc.        http://www.filtosh.com   
46026                         Putney      http://www.putneyvet.com   
30774                      Kite Ping       http://www.kiteping.com   
61057                  Valor Medical       http://valormedical.com   
57516    The Little Blue Book Mobile  http://thelittlebluebook.com   
9372                         Camiant        http://www.camiant.com   
26136                      Huitongda             http://www.htd.cn   
32206  Librestream Technologies Inc.    http://www.librestream.com   

                               company_category_list funding_total_usd  \
55429                                          Games       143025351.2   
49660     Audio|Enterprise Software|Meeting Software            160374   
20354             Big Data Analytics|Recruiting|SaaS             40000   
46026                                  Biotechnology          49600064   
30774                                            NaN                 -   
61057                                  Biotechnology           4080255   
57516  Doctors|Medical Professionals|Mobile Commerce          10000000   
9372                                     Web Hosting          16100000   
26136                                     E-Commerce                 -   
32206                            Enterprise Software           8592884   

          status company_country_code     company_city  funding_rounds  \
55429   acquired                  FIN         Helsinki               3   
49660  operating                  ESP           Madrid               1   
20354  operating                  USA     Santa Monica               1   
46026  operating                  USA         Portland               5   
30774  operating                  USA  California City               1   
61057  operating                  USA        San Diego               1   
57516   acquired                  USA           Darien               1   
9372    acquired                  USA      Marlborough               2   
26136  operating                  CHN          Nanjing               1   
32206  operating                  CAN         Winnipeg               1   

      founded_at first_funding_at last_funding_at time_before_first_fund  
55429 2010-05-01       2010-10-01      2013-04-18               153 days  
49660 2010-07-30       2011-11-30      2011-11-30               488 days  
20354 2013-07-11       2014-06-11      2014-06-11               335 days  
46026 2006-01-01       2009-11-02      2014-06-23              1401 days  
30774 2015-07-20       2015-10-03      2015-10-03                75 days  
61057 2007-01-01       2009-08-05      2009-08-05               947 days  
57516 1987-01-01       2010-12-07      2010-12-07              8741 days  
9372  2003-01-01       2005-03-29      2007-01-11               818 days  
26136 2008-01-01       2011-12-01      2011-12-01              1430 days  
32206 2003-01-01       2008-10-01      2008-10-01              2100 days

In [38]:
companies[companies.company_country_code.isnull() == False].sample(5)

company_permalink             company_name  \
51169           /organization/shine-security  Shine Technologies Corp   
38295  /organization/nature-s-next-beverages  Nature's Next Beverages   
43957                 /organization/plannify                 Plannify   
16991                /organization/easybring                   Nimber   
32656          /organization/liquidware-labs          LiquidWare Labs   

                    homepage_url  \
51169        http://getshine.com   
38295                        NaN   
43957        http://plannify.com   
16991      http://www.nimber.com   
32656  http://liquidwarelabs.com   

                                   company_category_list funding_total_usd  \
51169                         Advertising|Cyber|Security           3300000   
38295                                                NaN                 -   
43957                                        Hospitality            932785   
16991  Collaboration|Collaborative Consumption|Curate...            850000   
32656                                           Software           2776316   

          status company_country_code company_city  funding_rounds founded_at  \
51169  operating                  USA    Sunnyvale               2 2011-01-01   
38295  operating                  USA  Los Angeles               1 2013-12-15   
43957  operating                  BRA          Itá               1 2012-11-26   
16991  operating                  GBR       London               1 2010-10-01   
32656  operating                  USA     Palatine               2 2009-01-01   

      first_funding_at last_funding_at time_before_first_fund  
51169       2011-12-01      2012-10-01               334 days  
38295       2014-10-19      2014-10-19               308 days  
43957       2014-02-05      2014-02-05               436 days  
16991       2015-08-10      2015-08-10              1774 days  
32656       2009-04-17      2014-07-21               106 days

In [39]:
companies.company_country_code.unique()

array(['IND', 'USA', nan, 'CHN', 'HKG', 'CAN', 'CHL', 'GBR', 'FRA', 'AUS',
       'DNK', 'ROM', 'AUT', 'KOR', 'NLD', 'SWE', 'JPN', 'RUS', 'SGP',
       'NOR', 'COL', 'ESP', 'BEL', 'IRL', 'ITA', 'ISR', 'NZL', 'CZE',
       'DEU', 'CHE', 'BRA', 'BGR', 'SVN', 'JOR', 'HUN', 'BWA', 'NGA',
       'FIN', 'TUR', 'IDN', 'ARE', 'CYP', 'POL', 'CRI', 'PRT', 'ARG',
       'TWN', 'KHM', 'THA', 'SVK', 'UKR', 'LTU', 'ISL', 'ZAF', 'MEX',
       'VEN', 'URY', 'MUS', 'KEN', 'PHL', 'MNE', 'VNM', 'GHA', 'PSE',
       'MYS', 'PER', 'EGY', 'PAN', 'LVA', 'GGY', 'ALB', 'UGA', 'HRV',
       'EST', 'LBN', 'GRC', 'PAK', 'NPL', 'LUX', 'SAU', 'IRN', 'GTM',
       'BAH', 'ARM', 'BGD', 'MDA', 'SRB', 'TUN', 'ZWE', 'BRB', 'NIC',
       'TAN', 'PRI', 'TTO', 'BHR', 'CMR', 'BLR', 'CYM', 'BRN', 'ECU',
       'SLV', 'MLT', 'ZMB', 'MKD', 'GIB', 'TGO', 'LAO', 'BMU', 'HND',
       'MCO', 'MMR', 'KAZ', 'QAT', 'DOM', 'MAR', 'LIE', 'GEO', 'MOZ',
       'DZA', 'JAM', 'KNA', 'AZE', 'SEN', 'RWA', 'DMA', 'UZB', 'BLZ',
       'OMN', '

## Checking All `dtypes`

In [40]:
def datasetsDtypes(dfList, datasetNames, dtypesCol="dtypes_", pad=""):
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.dtypes.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[dtypesCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [41]:
datasetsDtypes(dfAll, lstCsvNames)

acquisitions dtypes_1               companies         dtypes_2  \
0        company_permalink   object       company_permalink           object   
1             company_name   object            company_name           object   
2    company_category_list   object            homepage_url           object   
3     company_country_code   object   company_category_list           object   
4       company_state_code   object       funding_total_usd           object   
5           company_region   object                  status           object   
6             company_city   object    company_country_code           object   
7       acquirer_permalink   object            company_city           object   
8            acquirer_name   object          funding_rounds            int64   
9   acquirer_category_list   object              founded_at   datetime64[ns]   
10   acquirer_country_code   object        first_funding_at   datetime64[ns]   
11     acquirer_state_code   object         last_funding_at   datetime64[ns]   
12         acquirer_region   object  time_before_first_fund  timedelta64[ns]   
13           acquirer_city   object                                     None   
14             acquired_at   object                                     None   
15          acquired_month   object                                     None   
16            price_amount  float64                                     None   
17     price_currency_code   object                                     None   

                investments dtypes_3                   rounds dtypes_4  
0         company_permalink   object        company_permalink   object  
1              company_name   object             company_name   object  
2     company_category_list   object    company_category_list   object  
3      company_country_code   object     company_country_code   object  
4        company_state_code   object       company_state_code   object  
5            company_region   object           company_region   object  
6              company_city   object             company_city   object  
7        investor_permalink   object  funding_round_permalink   object  
8             investor_name   object       funding_round_type   object  
9     investor_country_code   object       funding_round_code   object  
10      investor_state_code   object                funded_at   object  
11          investor_region   object        raised_amount_usd  float64  
12            investor_city   object                              None  
13  funding_round_permalink   object                              None  
14       funding_round_type   object                              None  
15       funding_round_code   object                              None  
16                funded_at   object                              None  
17        raised_amount_usd  float64                              None

In [42]:
#AISHA: VISUALIZE

In [43]:
print("Aisha is happy ^^")

Aisha is happy ^^


In [44]:
#Aisha stop

## `rounds` Dataset

In [45]:
dfAllCols.iloc[:, -2:]

investments                   rounds
0         company_permalink        company_permalink
1              company_name             company_name
2     company_category_list    company_category_list
3      company_country_code     company_country_code
4        company_state_code       company_state_code
5            company_region           company_region
6              company_city             company_city
7        investor_permalink  funding_round_permalink
8             investor_name       funding_round_type
9     investor_country_code       funding_round_code
10      investor_state_code                funded_at
11          investor_region        raised_amount_usd
12            investor_city                         
13  funding_round_permalink                         
14       funding_round_type                         
15       funding_round_code                         
16                funded_at                         
17        raised_amount_usd

In [46]:
invests.shape, rounds.shape

((168634, 18), (114948, 12))

In [47]:
invests.head()

company_permalink        company_name company_category_list  \
0             /organization/0-6-com             0-6.com           Curated Web   
1    /organization/004-technologies    004 Technologies              Software   
2  /organization/01games-technology  01Games Technology                 Games   
3              /organization/0xdata              H2O.ai             Analytics   
4              /organization/0xdata              H2O.ai             Analytics   

  company_country_code company_state_code         company_region  \
0                  CHN                 22                Beijing   
1                  USA                 IL  Springfield, Illinois   
2                  HKG                NaN              Hong Kong   
3                  USA                 CA            SF Bay Area   
4                  USA                 CA            SF Bay Area   

    company_city                    investor_permalink  \
0        Beijing     /organization/dt-capital-partners   
1      Champaign       /organization/venturecapital-de   
2      Hong Kong     /organization/cyberport-hong-kong   
3  Mountain View             /organization/capital-one   
4  Mountain View  /organization/nexus-venture-partners   

            investor_name investor_country_code investor_state_code  \
0     DT Capital Partners                   CHN                  23   
1   VCDE Venture Partners                   DEU                   5   
2     Cyberport Hong Kong                   HKG                 NaN   
3             Capital One                   USA                  VA   
4  Nexus Venture Partners                   USA                  CA   

    investor_region investor_city  \
0          Shanghai      Shanghai   
1         Frankfurt     Frankfurt   
2         Hong Kong     Hong Kong   
3  Washington, D.C.        Mclean   
4       SF Bay Area    Menlo Park   

                           funding_round_permalink funding_round_type  \
0  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
1  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
2  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
3  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
4  /funding-round/383a9bd2c04f7038bb543ccef5ba3eae               seed   

  funding_round_code  funded_at  raised_amount_usd  
0                  A  3/19/2008          2000000.0  
1                NaN  7/24/2014                NaN  
2                NaN   7/1/2014            41250.0  
3                  B  11/9/2015         20000000.0  
4                NaN  5/22/2013          3000000.0

In [48]:
rounds.head()

company_permalink            company_name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

                               company_category_list company_country_code  \
0                                              Media                  IND   
1  Application Platforms|Real Time|Social Network...                  USA   
2  Application Platforms|Real Time|Social Network...                  USA   
3                                  Apps|Games|Mobile                  NaN   
4                                        Curated Web                  CHN   

  company_state_code company_region   company_city  \
0                 16         Mumbai         Mumbai   
1                 DE     DE - Other  Delaware City   
2                 DE     DE - Other  Delaware City   
3                NaN            NaN            NaN   
4                 22        Beijing        Beijing   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B    1/5/2015         10000000.0  
1                  A  10/14/2014                NaN  
2                NaN    3/1/2014           700000.0  
3                  B   1/30/2014          3406878.0  
4                  A   3/19/2008          2000000.0

In [49]:
len(invests["company_name"].unique()), len(rounds["company_name"].unique())

(44577, 66102)

Since there are companies in `rounds` dataset that are not in `investments` dataset, <br>
therefore we should keep `rounds` dataset for further analysis.

## `companies` Dataset

In [50]:
companies.sample(5).T

54047  \
company_permalink                                /organization/sportlyzer   
company_name                                                   Sportlyzer   
homepage_url                                    http://www.sportlyzer.com   
company_category_list   Mobile Software Tools|Social Network Media|Sof...   
funding_total_usd                                                  506576   
status                                                          operating   
company_country_code                                                  EST   
company_city                                                        Tartu   
funding_rounds                                                          3   
founded_at                                            2009-07-09 00:00:00   
first_funding_at                                      2009-08-07 00:00:00   
last_funding_at                                       2012-05-30 00:00:00   
time_before_first_fund                                   29 days 00:00:00   

                                                        12486  \
company_permalink                  /organization/colt-defense   
company_name                                     Colt Defense   
homepage_url                                  http://colt.com   
company_category_list   Defense|Law Enforcement|Manufacturing   
funding_total_usd                                    33000000   
status                                              operating   
company_country_code                                      USA   
company_city                                    West Hartford   
funding_rounds                                              1   
founded_at                                1836-01-01 00:00:00   
first_funding_at                          2015-02-10 00:00:00   
last_funding_at                           2015-02-10 00:00:00   
time_before_first_fund                    65419 days 00:00:00   

                                                19149  \
company_permalink              /organization/everyart   
company_name                                 everyArt   
homepage_url                      http://everyart.com   
company_category_list                     Curated Web   
funding_total_usd                                   -   
status                                         closed   
company_country_code                              USA   
company_city                           Lucerne Valley   
funding_rounds                                      1   
founded_at              2008-05-30 20:10:41.984389376   
first_funding_at                  2012-03-01 00:00:00   
last_funding_at                   2012-03-01 00:00:00   
time_before_first_fund   1370 days 03:49:18.015610624   

                                                                    28678  \
company_permalink                                   /organization/invibox   
company_name                                                      Invibox   
homepage_url                                           http://Invibox.com   
company_category_list   App Stores|Email|Enterprise Software|Productiv...   
funding_total_usd                                                  300000   
status                                                             closed   
company_country_code                                                  NaN   
company_city                                                          NaN   
funding_rounds                                                          1   
founded_at                                            2013-04-04 00:00:00   
first_funding_at                                      2013-05-20 00:00:00   
last_funding_at                                       2013-05-20 00:00:00   
time_before_first_fund                                   46 days 00:00:00   

                                           28321  
company_permalink        /organization/intematix  
company_name                           Intematix  
homepage_url            http://www.intematix.co

In [51]:
companies.isnull().sum()

company_permalink            0
company_name                 0
homepage_url              5058
company_category_list     3147
funding_total_usd            0
status                       0
company_country_code      6956
company_city              8026
funding_rounds               0
founded_at                   1
first_funding_at             3
last_funding_at              0
time_before_first_fund       0
dtype: int64

we will need to remove the none values of the following columns <br>
based on the question that we are trying to gain insight for: <br>
1.	Can the factors that affect a start-up’s growth be determined?
    -	For this: We are initially interested in <br> `funding_total_usd`, `status`, `funding_rounds`, `first_funding_at`, and `last_funding_at`
    -   So we could remove the 24 null values of `first_funding_at` since they are so small <br>
        compared to the rest of the dataset
2.  Do taxes play an important factor in deciding a start-up’s success?

